clippy_lints/src/methods.rs

```diff
@@ -1562,6 +1586,31 @@ fn lint_map_unwrap_or(cx: &LateContext<'_, '_>, expr: &hir::Expr, map_args: &[hi
     }
 }
 
+/// lint use of `map().flatten()` for `Iterators`
+fn lint_map_flatten<'a, 'tcx>(
+    cx: &LateContext<'a, 'tcx>,
+    expr: &'tcx hir::Expr,
+    _map_args: &'tcx [hir::Expr],
+    _flatten_args: &'tcx [hir::Expr],
+) {
+    // lint if caller of `.map().flatten()` is an Iterator
+    if match_trait_method(cx, expr, &paths::ITERATOR) {
+        let msg = "called `map(..).flatten()` on an `Iterator`. \
+                   This is more succinctly expressed by calling `.flat_map(..)`";
+        let self_snippet = snippet(cx, _map_args[0].span, "..");
+        let func_snippet = snippet(cx, _map_args[1].span, "..");
+        let hint = format!("{0}.flat_map({1})", self_snippet, func_snippet);
+        span_lint_and_then(cx, MAP_FLATTEN, expr.span, msg, |db| {
+            db.span_suggestion_with_applicability(
+                expr.span,
+                "try using flat_map instead",
+                hint,
+                Applicability::MachineApplicable,
```

Comment:
If it doesn't make mess with code formatted like:
```
vec![5_i8; 6]
    .into_iter()
    .map(|x| 0..x)
    .flatten()
    .collect();
```
Then MachineApplicable should be fine.

Reply:
Is there a way to setup a test that checks machine applicability? I'm having some problems getting the clippy-driver to work for manual testing and I'm not sure if I'm just going about this wrong.

Reply:
I tried testing it locally by invoking `cargo run --bin clippy-driver -- -L ./target/debug/ ~/git/rust/playground/src/main.rs` with just a main function around the snippet you posted above and it didnt even mention the lint. And beyond that I'm not sure how to get rustfix to apply something from a locally compiled clippy-driver :/

I'll leave it alone for now and stop worrying about it.

Reply:
Oh, for local testing. Add a ui test and run it and check the generated output. https://github.com/rust-lang-nursery/rust-clippy/blob/master/CONTRIBUTING.md#running-test-suite

clippy-driver can't be run directly

lmk if you have questions!

Reply:
Ooh, I have a UI test that works and verifies the help prompt and the lint and thats all good. I was just trying to check if it would cleanly apply. 

Also the testing manually section directly below the one you linked misled me into thinking I could run clippy driver directly :(